lib/apiservers/engine/backends/commit.go

```diff
@@ -0,0 +1,472 @@
+		Parent:          vc.LayerID,
+		Author:          config.Author,
+		Comment:         config.Comment,
+		Created:         time.Now().UTC(),
+		Container:       container.ID,
+		ContainerConfig: *container.Config,
+		Architecture:    runtime.GOARCH,
+		OS:              runtime.GOOS,
+		DockerVersion:   version.DockerServerVersion,
+		Config:          newConfig,
+		Size:            lm.Size,
+	}
+
+	m, err := json.Marshal(meta)
+	if err != nil {
+		return InternalServerError(fmt.Sprintf("Failed to marshal image layer config: %s", err))
+	}
+	// layer metadata
+	lm.Meta = string(m)
+	lm.Image.Parent = vc.LayerID
+	lm.Image.Store = host
+	return nil
+}
+
+func downloadDiff(rc io.ReadCloser, containerID string, options imagec.Options) (*imagec.ImageWithMeta, error) {
+	defer trace.End(trace.Begin(containerID))
+
+	// generate random string as layer ID
+	layerID := stringid.GenerateRandomID()
+
+	tmpLayerFileName, diffIDSum, gzSum, err := compressDiffToTmpFile(rc, containerID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Cleanup function for the error case
+	defer func() {
+		if err != nil {
+			os.Remove(tmpLayerFileName)
+		}
+	}()
+
+	blobSum := digest.NewDigestFromBytes(digest.SHA256, gzSum)
+	log.Debugf("container %s blob sum: %s", containerID, blobSum.String())
+	diffID := digest.NewDigestFromBytes(digest.SHA256, diffIDSum)
+	log.Debugf("container %s diff id: %s", containerID, diffID.String())
+
+	layerFile, err := os.Open(string(tmpLayerFileName))
+	if err != nil {
+		return nil, err
+	}
+	defer layerFile.Close()
+
+	decompressed, err := gzip.NewReader(layerFile)
+	if err != nil {
+		return nil, err
+	}
+	defer decompressed.Close()
+
+	// get a tar reader
+	tr := tar.NewReader(decompressed)
+
+	// iterate through tar headers to get file sizes
+	var layerSize int64
+	for {
+		tarHeader, terr := tr.Next()
+		if terr == io.EOF {
+			break
+		}
+		if terr != nil {
+			err = terr
+			return nil, err
+		}
+		layerSize += tarHeader.Size
+	}
+
+	if layerSize == 0 {
+		diffID = digest.Digest(dockerLayer.DigestSHA256EmptyTar)
+	}
+	log.Debugf("container %s size: %d", containerID, layerSize)
+
+	// Ensure the parent directory exists
+	destination := path.Join(imagec.DestinationDirectory(options), layerID)
+	err = os.MkdirAll(destination, 0755) /* #nosec */
+	if err != nil {
+		return nil, err
+	}
+
+	// Move(rename) the temporary file to its final destination
+	err = os.Rename(string(tmpLayerFileName), path.Join(destination, layerID+".tar"))
+	if err != nil {
+		return nil, err
+	}
+
+	// layer metadata
+	lm := &imagec.ImageWithMeta{
+		Image: &models.Image{
+			ID: layerID,
+		},
+		DiffID: diffID.String(),
+		Layer: imagec.FSLayer{
+			BlobSum: blobSum.String(),
+		},
+		Size: layerSize,
+	}
+	return lm, nil
+}
+
+// compressDiffToTmpFile will write stream to temp file, and return temp file name and tar file checksum, compressed file checksum
+func compressDiffToTmpFile(rc io.ReadCloser, containerID string) (string, []byte, []byte, error) {
```

Comment:
so far, I feel that this method is just compressing tar file to a gz file, and could get some knowledge about the checksum, and file size.
the caller should be responsible for how to use those values. So I just renamed the variables.